### PR TITLE
Factor ImmutableArray<T> into many files

### PR DESCRIPTION
**Pull request by davidwrighton**
- Break immutable array (an enumerable, immutable, array) into a core component
- And a bunch of algorithm components which represents algorithms on the array
- And a bunch of components which represent creation of new related immutable arrays
- The purpose of this work is to make it possible to source include the core of immutable array into a component which cannot depend on high level apis such as LINQ, SortedDictionary, etc. (All api use must be satisfied by System.Private.CoreLib). The work to add this other component (System.Private.Reflection.Metadata.Ecma335 will be delivered via a separate PR)

### CODE REVIEW

**Comment by davidwrighton** [comment]
This is just the work for immutable array refactor that I need for System.Private.Reflection.Metadata.Ecma335. I'd like opinions on whether this is the right set of partial classes.  (I broke this apart via partial classes at the request of Ian)

File: src/System.Collections.Immutable/src/System.Collections.Immutable.csproj [hunk 1]
```diff
@@ -40,6 +40,14 @@
     <Compile Include="System\Collections\Immutable\ImmutableArray_1.Builder.cs" />
     <Compile Include="System\Collections\Immutable\ImmutableArray_1.Enumerator.cs" />
     <Compile Include="System\Collections\Immutable\ImmutableArray_1.cs" />
+    <Compile Include="System\Collections\Immutable\ImmutableArray_1.CollectionInterfaces.cs" />
+    <Compile Include="System\Collections\Immutable\ImmutableArray_1.IImmutableList.cs" />
+    <Compile Include="System\Collections\Immutable\ImmutableArray_1.IndexOf.cs" />
+    <Compile Include="System\Collections\Immutable\ImmutableArray_1.Insert.cs" />
+    <Compile Include="System\Collections\Immutable\ImmutableArray_1.OfType.cs" />
```
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System.Collections.Immutable.csproj]: This file contains a single method that is less than 10 LOC, so I'm not so sure it deserves its own file. Maybe move this back into `ImmutableArray_1.cs`

Comment (davidwrighton) [inline src/System.Collections.Immutable/src/System.Collections.Immutable.csproj]: That one method uses the OfType LINQ extension method, which is one of the apis that does not exist in System.Private.CoreLib

Comment (ianhays) [inline src/System.Collections.Immutable/src/System.Collections.Immutable.csproj]: It does seem a bit overkill to separate all of these methods in their own files like `Insert`, `Remove`, and `Sort`. Why not just have two files: one file with the stuff you want so you can build it without higher APIs, and one file with everything else?

**Comment by danmosemsft** [comment]
@davidwrighton is this good for a final CR now?

**Comment by davidwrighton** [comment]
Not quite, I'm going to respond to feedback today, and then push for a CR

**Comment by davidwrighton** [comment]
I'm now ready for CR

**Pull request merged**